Some small tweaks to Red.

xinminglai opened the pull request:
Hi, These 3 patches add method column-values, and some small fixes for Red::Driver::Pg

Thanks!

xinminglai committed:
Add column-values method to MetamodelX::Red::Model

xinminglai committed:
Additional types text/json/jsonb info for Driver::Pg

xinminglai committed:
Fix In operation for Driver::Pg

FCO commented:
It seems 2 tests are failing with these changes...

FCO commented on lib/Red/Driver/Pg.pm6:
```diff
@@ -34,6 +34,22 @@ multi method translate(Red::AST::Select $_, $context?, :$gambi where !*.defined)
     self.Red::Driver::CommonSQL::translate($_, $context, :gambi);
 }
 
+multi method translate(Red::AST::In $_, $context?) {
```
Isn't it common? Couldn't it be on CommonSQL?

xinminglai replied:
It's common, But It's strange that it works in SQLite.

FCO commented on lib/Red/Driver/Pg.pm6:
```diff
@@ -34,6 +34,22 @@ multi method translate(Red::AST::Select $_, $context?, :$gambi where !*.defined)
 }
 
+multi method translate(Red::AST::In $_, $context?) {
+    if .right.value ~~ Positional {
```
I mean something like:
```perl6
multi method translate(Red::AST::In $_ where .right.value ~~ Positional, $context?) {
```

FCO replied:
But, instead of testing the right's value type, it would be better to test if it's an Red::AST::Value and if its `.type` is Positional.

```perl6
multi method translate(Red::AST::In $_ where .right ~~ Red::AST::Value && .right.type ~~ Positional, $context?) {
```

FCO commented on lib/Red/Driver/Pg.pm6:
```diff
@@ -34,6 +34,22 @@ multi method translate(Red::AST::Select $_, $context?, :$gambi where !*.defined)
+            return "$lstr { .op } (SELECT 0 WHERE false)" => @lbind;
+        }
+
+        my $in-placeholder = '(' ~ (self.wildcard xx .right.value.elems).join(',') ~ ')';
```
If you fix one of my errors (https://github.com/FCO/Red/blob/master/lib/Red/Driver/CommonSQL.pm6#L521) changing `"?"` to `self.wildcard`, you could just use the `self.translate: .right, $context`

FCO commented on lib/Red/Driver/Pg.pm6:
```diff
@@ -111,6 +127,10 @@ multi method default-type-for(Red::Column $ where .attr.type ~~ Bool
 multi method default-type-for(Red::Column $                                                 --> Str:D) {"varchar(255)"}
 
+multi method type-by-name("text" --> "text") {}
+multi method type-by-name("json" --> "json") {}
```
Could you, please, add a test for json and jsonb?

xinminglai commented on lib/Red/Driver/Pg.pm6:
```diff
@@ -34,6 +34,22 @@ multi method translate(Red::AST::Select $_, $context?, :$gambi where !*.defined)
+        my ($lstr, @lbind) := do given self.translate: .left,  $context { .key, .value }
+
+        if .right.value.elems == 0 {
+            return "$lstr { .op } (SELECT 0 WHERE false)" => @lbind;
```
Because In Pg, the   `in ()`  will raise error, So, we should add placehoder statement to select "nothing".

FCO replied:
In that way, maybe we could do an:

```perl6
multi method translate(Red::AST::In $_ where .right ~~ Red::AST::Value && .right.type ~~ Positional && .right.elems, $context?) {
```

?

xinminglai committed:
Merge remote-tracking branch 'fco/master'

xinminglai committed:
Merge remote-tracking branch 'fco/master'

xinminglai committed:
Use self.wildcard to implement the In with @array values.

xinminglai committed:
$data structure fix for Postgres, so it behave consistently with SQLite

for SQLite, it re-fetches the row from the model by using LastInsertedRow
class, Then the `.row{@ids}` does the right thing all the time. But in
postgres, the $data will filled by `INSERT INTO table xxx RETURNING *` after
the RETURNING, we'll $data with column names with underscores.
So, in the pg case, We convert the $data structure according to the model, make
$data look like the same as fetched with LastInsertedRow class.

xinminglai committed:
Use data-copy to prevent modify of immutable exception.

FCO commented on t/20-in-pg.t:
```diff
@@ -0,0 +1,48 @@
+
+model Category is table<test_category> {
+    has Int $.id is serial;
+    has Int $.parent_id is column{ :references{ Category.id }, :nullable, };
```
This test is failing here. For `Red:api<2>` you need to set the `:model<>` for `is referencing` or `:model-name<>` in this case... and receive the model type as parameter:
```perl6
has Int $.parent_if is column{ :references{ .id }, :nullable };
```
